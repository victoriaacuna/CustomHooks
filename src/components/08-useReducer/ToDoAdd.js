import React from 'react';
import { useForm } from '../../hooks/useForm';

export const ToDoAdd = ({handleAddToDo}) => {

    const [{description}, handleInputChange, reset] = useForm({
        description: '',

    });

    const handleSubmit = (e) => {
        e.preventDefault();

        if(description.trim().length <= 1){
            return;
        }

        const newToDo  = {
            id: new Date().getTime(),
            desc: description,
            done: false
        };

        handleAddToDo(newToDo);
        reset();

        
    }

    return (
        <div>
            <h3>Agregar To Do</h3>
            <form onSubmit={handleSubmit}>
                <input
                    type="text"
                    name="description"
                    placeholder="Aprender..."
                    autoComplete="off"
                    className="form-control"
                    onChange={handleInputChange}
                    value={description}
                />
                <button 
                    className="btn btn-outline-primary mt-2 mb-5 btn-block"
                    type="submit"
                >
                        Agregar
                </button>
            </form>
            
        </div>
    )
}
