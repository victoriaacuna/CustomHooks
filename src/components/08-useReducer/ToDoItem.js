import React from 'react'

export const ToDoItem = ({toDo, index, handleDelete, handleToggle}) => {

    const {id, desc, done} = toDo;


    return (
        <li 
            key={toDo.id}
            className="list-group-item"
            style={{display: "flex"}}
        >
            <p 
                className={`${done ?'complete mt-3':'mt-3'}`}
                onClick={() => {handleToggle(id)}}
            >
                    {index+1}. {desc}
            </p>
            <button className="btn btn-danger ml-5" 
            onClick={() => handleDelete(id)}
            >Eliminar</button>
        </li>
    )
}
