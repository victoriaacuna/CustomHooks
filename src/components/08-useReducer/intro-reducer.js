const { logDOM } = require("@testing-library/react");


const initialState = [{
    id: 1,
    toDo: 'Comer pan',
    done: false
}];

const toDoReducer = (state = initialState, action) => {

    if(action?.type === 'agregar'){
        return [...state, action.payload]
    }
    return state;
};

let toDos = toDoReducer();
console.log(toDos);

const newToDo = {
    id: 2,
    toDo: 'Comer queso',
    done: false
};

const action = {
    type: 'agregar',
    payload: newToDo
}

toDos = toDoReducer(initialState, action);

console.log(toDos);