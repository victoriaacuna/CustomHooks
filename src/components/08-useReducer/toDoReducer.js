
export const todoReducer = ( state = [], action) => {
    switch (action.type) {
        case 'add':
            return [...state, action.payload];
        case 'delete':
            console.log(action.payload);
            return state.filter(todo => todo.id !== action.payload);
        case 'toggle':
            // return state.map(toDo => {
            //     if(toDo.id === action.payload){
            //         return {
            //             ...toDo,
            //             done: !toDo.done
            //         } 
            //     } else {
            //         return toDo;
            //     }
            // })
            return state.map( toDo => 
                (toDo.id===action.payload) ? {... toDo, done: !toDo.done} : toDo
            )
        default:
            return state;
    }
    
}