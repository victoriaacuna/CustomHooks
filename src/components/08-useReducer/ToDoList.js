import React from 'react'
import { ToDoItem } from './ToDoItem'

export const ToDoList = ({toDos, handleDelete, handleToggle}) => {
    return (
        <div>
            <ul className="list-group list-group-flush">
            {
                toDos.map( (toDo, index) => {
                    return (
                        <ToDoItem
                            key={toDo.id} 
                            toDo = {toDo}
                            index = {index}
                            handleDelete = {handleDelete}
                            handleToggle = {handleToggle}
                        />
                    )
                })
                
            }
                
            </ul>
        </div>
    )
}
