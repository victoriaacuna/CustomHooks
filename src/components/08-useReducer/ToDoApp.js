import React, { useEffect, useReducer } from 'react';
import { ToDoAdd } from './ToDoAdd';
// import { useForm } from '../../hooks/useForm';
import { ToDoList } from './ToDoList';
import { todoReducer } from './toDoReducer';


const init = () => { 

    return JSON.parse(localStorage.getItem('toDos')) || [];

    // return [{
    //     id: new Date().getTime(),
    //     desc: 'Aprender react',
    //     done: false
    // }]
}

export const ToDoApp = () => {

    const [toDos, dispatch] = useReducer(todoReducer, [], init);
    

    useEffect(() => {
        localStorage.setItem('toDos', JSON.stringify(toDos));
        return () => {
            
        }
    }, [toDos])

    // console.log(description);

    // const handleSubmit = (e) => {
    //     e.preventDefault();

    //     if(description.trim().length <= 1){
    //         return;
    //     }

    //     const newToDo  = {
    //         id: new Date().getTime(),
    //         desc: description,
    //         done: false
    //     };

        
    // }

    const handleDelete = (toDoID) => {

        console.log(toDoID);
        const action = {
            type: 'delete',
            payload: toDoID
        };
        dispatch(action);

    }

    const handleToggle = (toDoID) => {
        dispatch({
            type: 'toggle',
            payload: toDoID
        })
    }

    const handleAddToDo = (newToDo) => {
        const action = {
            type: 'add',
            payload: newToDo
        };

        dispatch(action);
    }

    return (
        <div>
            <h1>ToDo App: ({toDos.length})</h1>
            <hr/>

            <div>
                <ToDoAdd handleAddToDo={handleAddToDo}/>
            </div>

            <ToDoList 
                toDos={toDos} 
                handleDelete={handleDelete}
                handleToggle={handleToggle}
            />
        </div>
    )
}
