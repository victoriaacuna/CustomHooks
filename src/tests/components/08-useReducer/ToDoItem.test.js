import '@testing-library/jest-dom';
import {shallow} from 'enzyme';
import React from 'react';
import { ToDoItem } from '../../../components/08-useReducer/ToDoItem';
import { demoToDo } from '../../fixtures/demoToDos';

describe('Pruebas en el ToDoItem', () => {

    const handleDelete = jest.fn();
    const handleToggle = jest.fn();

    const wrapper = shallow(<ToDoItem
        toDo = {demoToDo[0]}
        index = {0}
        handleDelete = {handleDelete}
        handleToggle = {handleToggle}
    />);

    test('should match snapshot', () => {
        expect(wrapper).toMatchSnapshot();
    });

    test('should call handleDelete', () => {
        wrapper.find('button').simulate('click');
        expect(handleDelete).toHaveBeenCalledWith(demoToDo[0].id);
    });

    test('should call handleToggle', () => {
        wrapper.find('p').simulate('click');
        expect(handleToggle).toHaveBeenCalledWith(demoToDo[0].id);
    });

    test('should show text correctly', () => {
        expect(wrapper.find('p').text().trim()).toBe(`1. ${demoToDo[0].desc}`);
    });

    test('should not have complete as a class', () => {
        expect(wrapper.find('.complete').exists()).toBeFalsy()
    })
    
    
    
    
})
