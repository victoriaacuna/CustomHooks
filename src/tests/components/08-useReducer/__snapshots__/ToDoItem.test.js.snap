// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en el ToDoItem should match snapshot 1`] = `
<li
  className="list-group-item"
  key="1"
  style={
    Object {
      "display": "flex",
    }
  }
>
  <p
    className="mt-3"
    onClick={[Function]}
  >
    1
    . 
    Aprender React
  </p>
  <button
    className="btn btn-danger ml-5"
    onClick={[Function]}
  >
    Eliminar
  </button>
</li>
`;
